inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/config/filebased/ConfigurationDirectoriesWatcher.java

```diff
@@ -0,0 +1,112 @@
+import lombok.val;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.stereotype.Service;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.config.service.ActivationConfigCondition;
+import rocks.inspectit.oce.core.config.service.ActivationConfigConditionMet;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.nio.file.StandardWatchEventKinds.*;
+
+/**
+ * Service for automatically watching all registered {@link DirectoryPropertySource}s.
+ * If any changes are detected, the configuration gets reloaded into the {@link InspectitEnvironment}.
+ */
+@Service
+@Slf4j
+@ActivationConfigCondition("config.fileBased.watch")
+@Conditional(ActivationConfigConditionMet.class)
+public class ConfigurationDirectoriesWatcher {
+
+    @Autowired
+    InspectitEnvironment env;
+
+    private WatchService ws = null;
+
+    private Map<WatchKey, DirectoryPropertySource> watchKeyToPropertySourceMap = new HashMap<>();
+
+    @PostConstruct
+    private void init() {
+        log.info("Starting config directory watch service..");
+        try {
+            ws = FileSystems.getDefault().newWatchService();
+        } catch (IOException e) {
+            log.error("Unable to create watch service", e);
+            return;
+        }
+        env.readPropertySources(propertySources ->
+                propertySources.stream()
+                        .filter(ps -> ps instanceof DirectoryPropertySource)
+                        .map(ps -> (DirectoryPropertySource) ps)
+                        .forEach(ps -> {
+                            Path dir = ps.getRootDir();
+                            try {
+                                WatchKey key = dir.register(ws, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+                                watchKeyToPropertySourceMap.put(key, ps);
+                            } catch (IOException e) {
+                                log.error("Unable to register dir " + dir.toString() + " for watching", e);
+                            }
+                        })
+        );
+        if (watchKeyToPropertySourceMap.isEmpty()) {
+            log.info("Terminating config directory watch service as no valid directory was registered");
+            stopWatchService();
+        } else {
+            startWatcherThread();
+        }
+    }
+
+    @PreDestroy
+    private void stopWatchService() {
+        if (ws != null) {
+            try {
+                ws.close();
+            } catch (IOException e) {
+                log.error("Unable to close watch service", e);
+                return;
+            }
+        }
+    }
+
+    private void startWatcherThread() {
+        val watchingThread = new Thread(() -> {
```

Comment:
OK I get what you did, what I would try to avoid is uncontrolled creation of threads.. A much better way would be the `Executor` usage.. Now the Java8 does bring some default executor stuff, but this one is not a `ScheduledExecutorService`.. As we wound anyway need some scheduling for metrics collections, let's already prepare a field for that.. I see two options:

1. Use Spring scheduling stuff, so just have a `@Scheduled` annotation method (needs @EnableScheduling in the configuration). This approach also makes it easer to test later on..
2. Create our own Scheduled executor which you can auto-wire and use when needed.. Then you can submit runnable or callables and get back a Cancelable object to control if it has to be canceled..

What do you think?